src/hyp_map.rs

```diff
@@ -271,6 +302,71 @@ impl HypPageTable {
             r.restore();
         }
     }
+
+    // Return the virtual address of U-mode mapping slot `slot`.
+    fn umode_slot_va(slot: u64) -> Result<PageAddr<SupervisorVirt>, Error> {
+        if slot < UMODE_MAPPING_SLOTS {
+            // Unwrap okay: the result is dependent on constant that must be page aligned.
+            Ok(PageAddr::new(RawAddr::supervisor_virt(
+                UMODE_MAPPINGS_START + slot * UMODE_MAPPING_SLOT_SIZE,
+            ))
+            .unwrap())
+        } else {
+            Err(Error::InvalidSlot)
+        }
+    }
+
+    // Map a set of pages pinned for sharing with the hypervisor into a U-mode slot.
+    fn map_umode_slot(
+        &self,
+        slot: u64,
+        pages: &PinnedPages,
+        writable: bool,
+    ) -> Result<PageAddr<SupervisorVirt>, Error> {
+        let num_pages = pages.range().num_pages();
+        if num_pages > PageSize::num_4k_pages(UMODE_MAPPING_SLOT_SIZE) {
+            return Err(Error::OutOfMap);
+        }
+        let vaddr = Self::umode_slot_va(slot)?;
+        let mapper = self
+            .sv48
+            .map_range(vaddr, PageSize::Size4k, num_pages, &mut || {
+                self.pte_pages.borrow_mut().next()
+            })
+            .map_err(|_| Error::MapperCreationFailed)?;
+        let perms = if writable {
+            PteFieldBits::leaf_with_perms(PteLeafPerms::URW)
+        } else {
+            PteFieldBits::leaf_with_perms(PteLeafPerms::UR)
+        };
+        for (virt, phys) in vaddr
+            .iter_from()
+            .zip(pages.range().base().iter_from())
+            .take(num_pages as usize)
+        {
+            // Safe because pages are mapped in U-mode and cannot be accessed by the hypervisor directly.
```

Comment:
See comment below, too.

The safety comment of mapper is now in a bit of a gray area now that we have u-mode mappings. The consensus when I started having u-mode mappings was that mappings in userspace shouldn't be considered aliases for salus because salus doesn't have a 'normal' memory access to the memory (SUM bit).

It is true that this might map the same set of pinned pages twice in u-mode, and would break the safety model of u-mode. But if -- as comment below seems to suggest -- salus shouldn't mark unsafe functions that might make u-mode unsafe, then this comment here is fine.

So yes, I am contradicting myself with the safety comments, and it is a confusing matter that must be clarified.

Reply:
For u-mode, it'll have to trust s-mode not to be malicious or buggy, it has no choice.

For salus, what needs to be expressed in this safety comment is what guarantees that salus doesn't have access to the PinnedPages that are being mapped. How is it guaranteed that those same pages aren't in the sv48 map used by salus?

Reply:
Ah, thank you for this. Let me digest all of this. In a hand-wavy explanation, the pages are mapped directly from guest, but yes, I need make this clear in the function calls and safety requirement. I'll look at it and update the patch.